### ISSUE TITLE
When edditing an adjustment, the tax type displayed is not the one selected previously

### ISSUE THREAD

**Issue by RachL** [openfoodfoundation/openfoodnetwork#5422]
## Description

When you edit an adjustment of an order the tax rate displayed is not the tax rate that has been selected previously. 

## Steps to Reproduce
1. Add an adjustment to an order and select a tax type
2. Save
3. Edit the adjustment again, see that the tax rate displayed is not the previous one: often it is the "remove tax" tax type

## Animated Gif/Screenshot

In this screenshot the 20% tax is displayed but we can clearly see that the tax rate that was calculated to get to 0,05 cents was not 20%

![image.png](https://images.zenhubusercontent.com/595e669a61b4ca3b0a737135/7bf24b8b-4b92-45d0-9a50-7b5402e5d2a1)

## Severity

bug-s3: a feature is broken but there is a workaround

## Your Environment
* Version used: v2.9.8
* Browser name and version: Firefox 73
* Operating System and version (desktop or mobile): Desktop

**Comment by luisramos0** [comment]
there must be a problem in the form or controller setting the tax_rate_id used app/views/spree/admin/adjustments/_edit_form.html.haml and probably set in app/controllers/spree/admin/adjustments_controller.rb

I am tagging good-first-issue, should be easy for a dev with experience with rails.

**Comment by luchiago** [comment]
Can I work on this?

**Comment by luisramos0** [comment]
yes! thanks!!

**Comment by luchiago** [comment]
@luisramos0 I found two things
The tax rate that is been used during the editing is correct. However, there is a config that is changing one boolean variable of the selected tax rate.
```ruby
34: def compute_line_item(line_item)
 => 35:   if line_item.tax_category == rate.tax_category
    36:     if rate.included_in_price
    37:       deduced_total_by_rate(line_item.total, rate)
    38:     else
    39:       round_to_two_places(line_item.total * rate.amount)
    40:     end
    41:   else
    42:     0
    43:   end
    44: end
```

```ruby
[16] pry(#<Spree::Calculator::DefaultTax>)> rate
=> #<Spree::TaxRate id: 1, amount: #<BigDecimal:564e41ba0be0,'0.2E0',9(18)>, zone_id: 3, tax_category_id: 1, created_at: "2020-06-07 00:38:33", updated_at: "2020-06-07 00:50:35", included_in_price: true, name: "Teste", show_rate_in_label: true, deleted_at: nil>
[17] pry(#<Spree::Calculator::DefaultTax>)> TaxRate.find_by_id(line_item.tax_category_id)
=> #<Spree::TaxRate id: 1, amount: #<BigDecimal:564e3ea59528,'0.2E0',9(18)>, zone_id: 3, tax_category_id: 1, created_at: "2020-06-07 00:38:33", updated_at: "2020-06-07 00:50:35", included_in_price: false, name: "Teste", show_rate_in_label: true, deleted_at: nil>
```

As you can se, the rate is the selected tax rate (in the case, 20%), but the attribute `included_in_price` was changed, causing the incorrect value

My guess is this configuration on this method from `tax_rate_decorator.rb`  used on `adjustments_controller.rb`

```ruby
    def compute_tax(amount)
      line_item = LineItem.new quantity: 1
      line_item.tax_category = tax_category
      line_item.define_singleton_method(:price) { amount }

      # Tax on adjustments (represented by the included_tax field) is always inclusive of
      # tax. However, there's nothing to stop an admin from setting one up with a tax rate
      # that's marked as not inclusive of tax, and that would result in the DefaultTax
      # calculator generating a slightly incorrect value. Therefore, we treat the tax
      # rate as inclusive of tax for the calculations below, regardless of its original
      # setting.
      with_tax_included_in_price do
        calculator.compute line_item
      end
    end

    private

    def with_tax_included_in_price
      old_included_in_price = included_in_price

      self.included_in_price = true
      calculator.calculable.included_in_price = true

      result = yield
    ensure
      self.included_in_price = old_included_in_price
      calculator.calculable.included_in_price = old_included_in_price

      result
    end
  end
```

What do you think?

**Comment by luisramos0** [comment]
I dont think there's an incorrect value, do you?
The fix to this issue should just change the drop down selection to the correct tax rate.
Maybe just something wrong with this line?
      = select_tag :tax_rate_id, options_from_collection_for_select(Spree::TaxRate.all, :id, :name, @tax_rate_id), prompt: t(:remove_tax), class: 'select2 fullwidth'

**Comment by luchiago** [comment]
Hmmm, I will check this @luisramos0

**Comment by luchiago** [comment]
@luisramos0 Even if I change the displayed tax, the first calculation (when the user selects a new tax rate) isn't correct. Correct me if I am wrong, but the correct value of the tax in the [screenshot](https://camo.githubusercontent.com/d5b603d8cae7d541d4ac0aa9f34b374a4ccf753c/68747470733a2f2f696d616765732e7a656e68756275736572636f6e74656e742e636f6d2f3539356536363961363162346361336230613733373133352f37626632346238622d346239322d343564302d396135302d376235343032653564326131) shouldn't be `1.83 * 0.2 = 0.36`?

**Comment by luisramos0** [comment]
I think you want to fix the wrong part of the equation.
I think the value is correct, the tax rate 20% is wrong because the select box is not picking up the correct tax rate.

**Comment by luchiago** [comment]
@luisramos0 I tried to change in the views but I'm having a hard time with this issue. There are two methods inside `app/controllers/spree/admin/adjustments_controller.rb` that I couldn't understand what they do and couldn't test them.

**Comment by luisramos0** [comment]
Those methods
@tax_rate_id = tax_rate_with_matching_tax || tax_rate_valid_for_order
they are looking for the correct tax rate for the adjustment. I wonder why do we need to look for it and not just fetch it through some association from the DB...

You can try to debug, do you get this error triggered?
@adjustment.errors.add :tax_rate_id, I18n.t(:adjustments_tax_rate_error)

This iss probably what is happening here, the code cant find the right tax rate and is just showing the first one on the list "tax_rate_valid_for_order".
One of the problem with this approach is that, if tax rates are changed after the order is placed, this look up will fail and we loose track of the rate that was applied to the adjustment...
This could be one explanation for the issue and the solution is not obvious.

I am removing the good first issue tag now that I understand what's going on.

**Comment by luisramos0** [comment]
@RachL this is now looking like a good candidate for the "tax overhaul" epic.

**Comment by RachL** [comment]
😭

**Comment by sigmundpetersen** [comment]
@luchiago do you still want to handle this? or do you rather want to find another issue to work on? 🙂

**Comment by luchiago** [comment]
@sigmundpetersen honestly, I don't know anymore where should I start with this issue :sweat_smile: but I still want to look forward with some help.

**Comment by RachL** [comment]
hello @luchiago thank you so much for everything you have contributed so far! I'm afraid our dev team has a lot on their plate currently and won't be able to help you with this one for now. So I'm unassigning you but don't hesitate to ping us if you feel you could move forward :)

**Comment by filipefurtad0** [comment]
Hey @RachL,
after PR #7112 I can't seem to reproduce this issue - looks fixed - (see vid below):

![Peek 2021-03-15 15-28](https://user-images.githubusercontent.com/49817236/111178799-66746680-85a3-11eb-9fa8-17359c4f4773.gif)

**Comment by RachL** [comment]
@filipefurtad0 I see you are selecting the Fulll VAT rate, but when you come back to editing the adjustment, the new_tax_rate is there. So it does not look fixed?But maybe I'm lost in the gif :)

**Comment by filipefurtad0** [comment]
Good point -  tax rate displayed is still not the previously selected one.

I've tried this a bit and it does work in several cases, for some tax rates. I ended up choosing a bad gif to demonstrate this, but yeah - it demonstrates that it is still an open issue :+1:
